modules/light-clients/08-wasm/light_client_module_test.go

```diff
@@ -77,209 +186,1378 @@ func (suite *WasmTestSuite) TestRecoverClient() {
+			suite.Require().True(found)
+
+			tc.malleate()
+
+			updateFunc := func() {
+				lightClientModule.UpdateStateOnMisbehaviour(suite.chainA.GetContext(), clientID, clientMsg)
+			}
+
+			if tc.panicErr == nil {
+				updateFunc()
+				if expectedClientStateBz != nil {
+					store := suite.chainA.App.GetIBCKeeper().ClientKeeper.ClientStore(suite.chainA.GetContext(), endpoint.ClientID)
+					suite.Require().Equal(expectedClientStateBz, store.Get(host.ClientStateKey()))
+				}
+			} else {
+				suite.Require().PanicsWithError(tc.panicErr.Error(), updateFunc)
+			}
+		})
+	}
+}
+
+func (suite *WasmTestSuite) TestRecoverClient() {
+	var (
+		expectedClientStateBz               []byte
+		subjectClientID, substituteClientID string
+	)
+
+	testCases := []struct {
+		name     string
+		malleate func()
+		expErr   error
+	}{
+		{
+			"success",
+			func() {
+				suite.mockVM.RegisterSudoCallback(
+					types.MigrateClientStoreMsg{},
+					func(_ wasmvm.Checksum, _ wasmvmtypes.Env, sudoMsg []byte, store wasmvm.KVStore, _ wasmvm.GoAPI, _ wasmvm.Querier, _ wasmvm.GasMeter, _ uint64, _ wasmvmtypes.UFraction) (*wasmvmtypes.ContractResult, uint64, error) {
+						var payload types.SudoMsg
+						err := json.Unmarshal(sudoMsg, &payload)
+						suite.Require().NoError(err)
+
+						suite.Require().NotNil(payload.MigrateClientStore)
+						suite.Require().Nil(payload.UpdateState)
+						suite.Require().Nil(payload.UpdateStateOnMisbehaviour)
+						suite.Require().Nil(payload.VerifyMembership)
+						suite.Require().Nil(payload.VerifyNonMembership)
+						suite.Require().Nil(payload.VerifyUpgradeAndUpdateState)
+
+						bz, err := json.Marshal(types.EmptyResult{})
+						suite.Require().NoError(err)
+
+						// TODO(jim): re-arrange export_test.go contents
+						prefixedKey := []byte("subject/")
+						prefixedKey = append(prefixedKey, host.ClientStateKey()...)
+						expectedClientStateBz = wasmtesting.CreateMockClientStateBz(suite.chainA.Codec, suite.checksum)
+						store.Set(prefixedKey, expectedClientStateBz)
+
+						return &wasmvmtypes.ContractResult{Ok: &wasmvmtypes.Response{Data: bz}}, wasmtesting.DefaultGasUsed, nil
+					},
+				)
+			},
+			nil,
+		},
+		{
+			"failure: cannot parse malformed substitute client ID",
+			func() {
+				substituteClientID = ibctesting.InvalidID
+			},
+			host.ErrInvalidID,
+		},
+		{
+			"failure: substitute client ID does not contain 08-wasm prefix",
+			func() {
+				substituteClientID = tmClientID
+			},
+			clienttypes.ErrInvalidClientType,
+		},
+		{
+			"failure: cannot find subject client state",
+			func() {
+				subjectClientID = unusedWasmClientID
+			},
+			clienttypes.ErrClientNotFound,
+		},
+		{
+			"failure: cannot find substitute client state",
+			func() {
+				substituteClientID = unusedWasmClientID
+			},
+			clienttypes.ErrClientNotFound,
+		},
+		/* TODO(jim): We don't pass a client state directly now so we need to modify how this is tested.
+		{
+			"failure: invalid substitute client state",
```

Comment:
Maybe this could be tested if we pass a `substituteClientID` for  a non-08-wasm client, and then the casting [here](https://github.com/cosmos/ibc-go/blob/527ce1d37f918ae4c16a1685191f1772800710bb/modules/light-clients/08-wasm/types/proposal_handle.go#L20) would fail...